File: connectors/whatsapp-web/src/media-storage.ts
/**
 * MinIO uploader for WhatsApp media. Bucket `socialmedia-media` is created
 * out-of-band; we just upload here. Object keys follow `attachments/{messageId}/{ts}.{ext}`.
 *
 * On startup the connector calls `ensureMediaBucket()` to make idempotent the bucket
 * existence check (cheap — single statObject call).
 */
import * as MinIO from 'minio';
import * as fs from 'fs';
import * as https from 'https';

const MINIO_ENDPOINT = process.env.MINIO_ENDPOINT || 'minio:9000';
const MINIO_ACCESS_KEY = process.env.MINIO_ACCESS_KEY || 'minioadmin';
const MINIO_SECRET_KEY = process.env.MINIO_SECRET_KEY || 'minioadmin';
const MINIO_USE_SSL = (process.env.MINIO_USE_SSL || 'true').toLowerCase() === 'true';
const MINIO_CA_CERT = process.env.MINIO_CA_CERT;
const BUCKET = process.env.MINIO_BUCKET || 'socialmedia-media';

const [host, portStr] = MINIO_ENDPOINT.split(':');
const port = portStr ? parseInt(portStr, 10) : MINIO_USE_SSL ? 443 : 9000;

const opts: MinIO.ClientOptions = {
  endPoint: host,
  port,
  useSSL: MINIO_USE_SSL,
  accessKey: MINIO_ACCESS_KEY,
  secretKey: MINIO_SECRET_KEY,
};
if (MINIO_USE_SSL && MINIO_CA_CERT && fs.existsSync(MINIO_CA_CERT)) {
  const ca = fs.readFileSync(MINIO_CA_CERT, 'utf-8');
  opts.transportAgent = new https.Agent({ ca, rejectUnauthorized: true });
}

const client = new MinIO.Client(opts);

const EXT_BY_MIME: Record<string, string> = {
  'image/jpeg': 'jpg',
  'image/png': 'png',
  'image/gif': 'gif',
  'image/webp': 'webp',
  'video/mp4': 'mp4',
  'video/quicktime': 'mov',
  'video/webm': 'webm',
  'audio/mpeg': 'mp3',
  'audio/ogg': 'ogg',
  'audio/wav': 'wav',
  'audio/mp4': 'm4a',
  'application/pdf': 'pdf',
};

function pickExt(mimeType: string | undefined, fileName: string | undefined): string {
  if (fileName && fileName.includes('.')) {
    return fileName.split('.').pop()!.slice(0, 6);
  }
  if (mimeType && EXT_BY_MIME[mimeType]) return EXT_BY_MIME[mimeType];
  if (mimeType?.startsWith('image/')) return 'bin';
  return 'bin';
}

export async function ensureMediaBucket(): Promise<void> {
  const exists = await client.bucketExists(BUCKET).catch(() => false);
  if (!exists) {
    await client.makeBucket(BUCKET, 'us-east-1');
  }
}

export interface UploadedMedia {
  storageKey: string;
  fileSize: number;
}

export async function uploadMedia(
  messageId: bigint | number,
  data: Buffer,
  mimeType: string | undefined,
  fileName: string | undefined
): Promise<UploadedMedia> {
  const ts = Date.now();
  const ext = pickExt(mimeType, fileName);
  const storageKey = `attachments/${messageId}/${ts}.${ext}`;
  const meta: Record<string, string> = {};
  if (mimeType) meta['Content-Type'] = mimeType;
  if (fileName) meta['x-amz-meta-filename'] = fileName;
  await client.putObject(BUCKET, storageKey, data, data.length, meta);
  return { storageKey, fileSize: data.length };
}

File: connectors/telegram-sync/sync/media_storage.py
"""MinIO uploader for Telegram media — used by history.py + realtime.py + backfill.

Bucket `socialmedia-media` is shared across WA + TG + IG. Object keys follow
`attachments/{message_id}/{ts}.{ext}` so they don't collide. We upload bytes
directly from BytesIO (no disk write).
"""
from __future__ import annotations

import io
import logging
import os
import time
from typing import Optional

from minio import Minio

logger = logging.getLogger(__name__)

MINIO_ENDPOINT = os.environ.get("MINIO_ENDPOINT", "minio:9000")
MINIO_ACCESS_KEY = os.environ.get("MINIO_ACCESS_KEY", "minioadmin")
MINIO_SECRET_KEY = os.environ.get("MINIO_SECRET_KEY", "minioadmin")
MINIO_USE_SSL = os.environ.get("MINIO_USE_SSL", "true").lower() == "true"
MINIO_BUCKET = os.environ.get("MINIO_BUCKET", "socialmedia-media")
# Use cert_check=False to accept the cluster's self-signed CA. The CA file is
# at /certs/ca.crt inside the container; for a stricter setup use
# `Minio(..., http_client=urllib3.PoolManager(ca_certs="/certs/ca.crt"))`.
_CERT_CHECK = os.environ.get("MINIO_CERT_CHECK", "false").lower() == "true"

_client: Optional[Minio] = None


def _get_client() -> Minio:
    global _client
    if _client is None:
        _client = Minio(
            MINIO_ENDPOINT,
            access_key=MINIO_ACCESS_KEY,
            secret_key=MINIO_SECRET_KEY,
            secure=MINIO_USE_SSL,
            cert_check=_CERT_CHECK,
        )
    return _client


def ensure_bucket() -> None:
    c = _get_client()
    if not c.bucket_exists(MINIO_BUCKET):
        c.make_bucket(MINIO_BUCKET)
        logger.info(f"Created MinIO bucket: {MINIO_BUCKET}")


_EXT_BY_MIME = {
    "image/jpeg": "jpg",
    "image/png": "png",
    "image/gif": "gif",
    "image/webp": "webp",
    "video/mp4": "mp4",
    "video/quicktime": "mov",
    "audio/ogg": "ogg",
    "audio/mpeg": "mp3",
    "audio/mp4": "m4a",
    "audio/wav": "wav",
    "application/pdf": "pdf",
}


def pick_ext(mime_type: Optional[str], file_name: Optional[str]) -> str:
    if file_name and "." in file_name:
        return file_name.rsplit(".", 1)[-1][:6].lower()
    if mime_type and mime_type in _EXT_BY_MIME:
        return _EXT_BY_MIME[mime_type]
    if mime_type and mime_type.startswith("image/"):
        return "bin"
    return "bin"


def upload_media(
    message_id: int,
    data: bytes,
    mime_type: Optional[str],
    file_name: Optional[str],
) -> tuple[str, int]:
    """Upload bytes to MinIO and return (storage_key, size). Storage key goes
    into attachments.file_url; the dashboard streams via /api/messages/media-blob/{id}."""
    c = _get_client()
    ext = pick_ext(mime_type, file_name)
    storage_key = f"attachments/{message_id}/{int(time.time() * 1000)}.{ext}"
    buf = io.BytesIO(data)
    metadata = {}
    if mime_type:
        metadata["Content-Type"] = mime_type
    c.put_object(
        MINIO_BUCKET,
        storage_key,
        buf,
        length=len(data),
        content_type=mime_type or "application/octet-stream",
        metadata=metadata or None,
    )
    return storage_key, len(data)
